### ISSUE TITLE
Support for multiple regions in GridSelection and colors for each

### ISSUE THREAD

**Issue by BrianHung** [glideapps/glide-data-grid#141]
`glide-data-grid` currently supports cell selection, range selection, row selection, and column selection.

```ts
// https://github.com/glideapps/glide-data-grid/blob/main/packages/core/src/data-grid/data-grid-types.ts
export interface GridSelection {
    readonly cell: readonly [number, number]; // cell selection
    readonly range: Readonly<Rectangle>; // column selection
}

export class CompactSelection {
  ...


}

// https://github.com/glideapps/glide-data-grid/blob/main/packages/core/src/data-grid/data-grid.tsx
readonly selectedRows?: CompactSelection;
readonly selectedColumns?: CompactSelection;
readonly selectedCell?: GridSelection;

```

However, those selections given as props are limited when representing selections that originate from interactions outside of the normal current grid and when customizing the colors and fill of individual selections. 

As an example / application from google sheets, the selections come from the active input box, and can be all different types and colors.

![image](https://user-images.githubusercontent.com/15721634/151451559-6e5595eb-1ab6-4a35-8ade-47ab23b64be0.png)

Would it possible for this library to support that? As a quick sketch, I think it could look like this

```ts
readonly selectedRows?: CompactSelection;
readonly selectedColumns?: CompactSelection;
readonly selectedCell?: GridSelection;
readonly selections?: ((CompactSelection | GridSelection) & SelectionStyle)[];

export interface SelectionStyle {
  border?: string // css equivalent
  backgroundColor?: string // css equivalent
}

```

And then `selections` would only need to be passed as a prop into `drawGrid`.

**Comment by pzcfg** [comment]
It wouldn't get you the borders, but you could do the background colors yourself by setting the cell's `bgCell` in `themeOverride`

https://github.com/glideapps/glide-data-grid/blob/a29d0ba1dbf7f6f69857bb13655937251bd5b680/packages/core/src/data-grid/data-grid-types.ts#L240-L245

**Comment by BrianHung** [comment]
Thanks for the pointer! The following type would be a better and more extensive styling.

```ts
readonly selections?: ((CompactSelection | GridSelection) & Partial<Theme>)[];
```

https://github.com/glideapps/glide-data-grid/blob/a29d0ba1dbf7f6f69857bb13655937251bd5b680/packages/core/src/common/styles.ts#L3-L42

**Comment by jassmith** [comment]
How would you want the keyboard + mouse interactions to work?

**Comment by BrianHung** [comment]
Keyboard + mouse interactions with `selections` can be left up to end user since `onKeyDown` and `onMouseDown` are already available props that could be used to update `selections` and override default behavior.

**Comment by jassmith** [comment]
Turns out I think we need this for other reasons as well. To support "transparent" features like sorting and grouping without requiring the user to implement them fully or partially we will need to be able to support multiple ranges as a seleciton.

This way if a table is sorted to the user a rectangular selection in the sorted table can be broken back out to its constituent sub-regions when reporting up to the consuming application without them having to consider what the sorting was doing. It's a complicated feature set but one that will help a lot not just with your use case but with many others.

**Comment by jassmith** [comment]
## Goals

Allow the user to highlight multiple areas of the grid to indicate related cells and regions.

## Implementation Details

```ts
export interface GridSelection {
    readonly cell: readonly [number, number];
    readonly range: Readonly<Rectangle> | readonly Readonly<Rectangle>[];
    readonly highlights: {
        readonly range: readonly Readonly<Rectangle>;
        readonly color?: string;
        readonly key? string;
    }[];
}
```

There are a variety of subtleties to this that I think are worth discussing.

#### Why not just use a single array of ranges?
The range in the GridSelection is basically a special highlight. It is the highlight which can be mutated via the keyboard. If you press shift+arrow keys it is the highlight which changes. If you press copy, it is the highlight whose contents are to be copied. It is also the only highlight which must contain the selected cell.

#### How does the primary range/selection get a color?
From the theme like it does now. The others get passed colors.

#### What about remapped ranges?
This might be a pipe dream to make this totally transparent. At the moment my thinking is that the sorting hook can also provide `getCellContent`, `getCellsForSelection`, as well as handle all events to deal with selection. Selection then needs to be mangled/demangled based on the sort index map.

I don't immediately see why it won't work, but I didn't see why column sorting couldn't be totally transparent until I tried to implement it either.

#### Is this backwards compatible?
Kind of? At the very least all existing implementations will continue to work. Any library which is based on an older version of the code prior to this change may be incompatible with any code that uses the newly expanded type however. Effectively the answer is yes for applications, no for libraries.

#### Tags
I think it would be very useful for each highlight region to have a tag. If either the top or the bottom of the region (prefer top) is visible, we could draw a tag labeling. Problem here is I need to think very carefully about how this would impact blitting. At the bare minimum we would need to be careful when doing a damage draw that is right below one of these highlights.

#### Do highlight regions also need to support lists?

Technically to make these work with transparent sorting we would need to allow them to be broken up as well. However this is a feature that is only needed in very specific cases and I'm not sure we don't just disable these whenever temporary sorting/filtering is enabled. Conversely these are in fact consumption only as far as DataEditor is concerned so breaking them up simply by creating a much larger list where some of them have the same color and making sure that we merge nearby ranges with the same color by treating the color as a key would also be sufficient.

This has the unfortunately side effect of slightly leaking this behavior if a developer were to add two highlight ranges next to each other which have the same associate color. To resolve this we would need to force a key parameter as well. It could be optional and transparent transform hooks could set it if its not already set. The key would allow for a hard line between adjacent highlights with the same color. Anything without a key will be assigned a random key during normalization.

## Issues

- No method for providing a mechanism to support drawing dashed lines.
- How does this interact with multiple overlapping highlights? Do the colors blend? What blend mode? Also ew...
- If two highlight regions are overlapping how do they outlines get their colors? They can't really blend.
- Can we draw this performantly? I suspect yes but I'm not 100% sure.

## Version bump requirements

Bare minimum minor version bump. There is a reasonable argument to be made for major version.

**Comment by jassmith** [comment]
It might be worth breaking this up into 2 work items

1) Support array of ranges
2) Support highlights

**Comment by BrianHung** [comment]
> No method for providing a mechanism to support drawing dashed lines.

A `highlight` could have a `drawRange` callback as a prop and which passes canvas `ctx` as an argument. Something akin to this I've seen in another library is https://prosemirror.net/docs/ref/#view.Decoration^widget which passes current state and positioning.

> How does this interact with multiple overlapping highlights? Do the colors blend? What blend mode? Also ew...

Allow that to be customized by end-user for each `highlight` as mentioned before via `drawRange`.

> If two highlight regions are overlapping how do they outlines get their colors? They can't really blend.

`highlights` should operate as a stack here: an element at end of array is last drawn, so visually on-top of previous elements. So new highlighting ranges (typically pushed to end of array) would always take precedence.

If end-user has problem with this behavior, they could just re-arrange the array before passing it into `data-grid`.

**Comment by jassmith** [comment]
I've got the basic works of this up in a branch. The colored highlight ranges aren't in yet but that will come next.

**Comment by jassmith** [comment]
Fixed in 4.0.0
